### ISSUE TITLE
Country cards.

### ISSUE THREAD

**Issue by santisis** [wfx/teg#3]
When you don't own a country and you get the card of that country for the first time when you conquer that country you be able to put armies in that country. That's OK.

If you already have win armies for one country, changed that card with other two to get more armies, you owns that country and you get the same country card again you can put armies in that country. That's OK too.

Now, if you have played the card for one country, then you get the same card again but you don't own that country when you conquer it you do not have choice to put armies in that country. That's wrong.

I think that the game isn't cleaning a boolean registry when you change the three cards for armies in order to let you put armies in a country if you get the same card again. It's only working if you own the country at the moment you get the card.

**Comment by eternautaCAT** [comment]
Thanks for the report, you are right, that is a bug. I will try to fix that, the only problem is that is a bug that is difficult to replicate and test, but i agree with you that can be only a boolean registry uncleaned.

**Comment by eternautaCAT** [comment]
I revised the code, and all seems right. What do you mean with "played the card for one country"? Can you explain better the situation? Thank you!

**Comment by santisis** [comment]
Perdón, recién lo leo hoy. Soy bastante malo expresándome en inglés, así que si encima no se entendió de una voy en castellano...

La secuencia es esta:
1) Recibo carta para algún país por primera vez.
2) Ocupo ese país (o lo tenía ocupado desde antes, es indistinto) => El sistema me deja poner tres fichitas en el país.
3) Pierdo ese país.
4) Vuelvo a obtener la tarjeta de ese país.
5) Ocupo ese país => El sistema no me deja poner fichitas (con "sistema" me refiero a la interfaz gráfica del juego).

Si no existiera el paso 3), o sea si no perdiera nunca el país, cuando en el paso 4) obtengo la tarjeta otra vez el sistema automáticamente me coloca las fichitas.

Entiendo que el comportamiento esperado es el que se da cuando no perdés el país. Podés hacer canje de la tarjeta de ese país tantas veces como la tarjeta te toque en suerte después de hacer los canjes de ejércitos y volverla al mazo. Funciona bien cuando tenés el país no funciona cuando lo tenés que conquistar.

Juego (bastante seguido) en Debian, en la versión empaquetada. El bug es reproducible cada vez que se da la secuencia. No sé si puedo ayudarte con algún log o más información.

**Comment by santisis** [comment]
Ah, aclaro que sé programar, si me tirás una punta de por dónde buscar en el fuente puedo intentar dar una mano.

**Comment by eternautaCAT** [comment]
No pienso que sea un bug. Fijate el reglamento, en la página 8, es claro de que el premio se recibe por única vez. Por qué perder le pais y luego recuperarlo debiera hacer que cambie el "por única vez"?

http://openpool.com.ar/reglamento/T.E.G.-REGLAMENTO.pdf

**Comment by santisis** [comment]
(Like everything in TEG is a matter of opinion. My interpretation is "por única vez mientras conserves la tarjeta".)

Beyond the opinion, now the behavior is inconsistent. When you get the card the system doesn't check if you have used it before.

Searching in the source code, look at the file server/play.c:823-832. It only checks if the player owns the country, not if the player has owned and played that card before.

(I repeat: I think that play.c:823-832 is the correct behavior and the other is the broken one.)

**Comment by santisis** [comment]
Me again ;).
I'm reading the sources and I can't understand the behavior of the cards stuff.

1) I didn't seen a persistence of the counties in the user. So, as the ADT are implemented the "por única vez" only can be as "por única vez en todo el juego, aunque la tarjeta le toque a otro después" ;). (On the other hand, the rules says that the cards are in a deck, in a real game cards doesn't repeat so often.)

2) The status of a country and a card are stored in g_countries[]. The member tarjeta is a ADT TARJETA wich contains the members numjug and usada.

3) The source fragments that interact with the tarjeta are:
   a) common/country.c:127-129 (lines 128-129 are useless and breaks the type abstraction), when country initializes, tarjeta is set to not usada.
   b) server/play.c:1013-1015, when a exchange is done, tarjeta is set to not usada.
   c) server/play.c:949 and sever/play.c:832, when a player puts the two reinforcements in the country, tarjeta is set to usada.

I could not find a place where the status of used in the card isn't resetting when the card is returned to the deck. Reading the source code of the server looks like if you receive the same card twice you could put the reinforcements every time... but playing with the client the option is not available. The reinforcements are done only when you owns the country, where the server does the change automatically.

As the server does seem to be handling the change correctly, could it be a bug in the client?

**Comment by santisis** [comment]
Bingo!

Reading client side of implementation, client/canje.c:39, it does the change of cards but never reset the internal status of that card, so, when client/gui-gnome/cards.c:455-481 test if card is available to do the reinforcements thing fails to enable the button.

The patch is something like adding (I'm not testing ir, only reading the sourcecode from GitHub)
tarjeta_poner(&g_countries[p1].tarjeta);
tarjeta_poner(&g_countries[p2].tarjeta);
tarjeta_poner(&g_countries[p3].tarjeta);
near the line 39 of client/canje.c.

**Comment by wfx** [comment]
Can you test and send a patch if it works?

Thanks,
wfx 

---- santisis schrieb ----

**Comment by santisis** [comment]
OK. I've test the suggested patch and it seems to work. I don't known if the place where I fixed the bug is the best place to do it.

Fist I've hacked the server replacing the implementation of get_random_country() at helper.c with:

PCOUNTRY get_random_country( get_random_func func )
{
        int i;
        for(i = 0; i < COUNTRIES_CANT; i++)
                if(func(i))
                        return &g_countries[i];
}

So the server returns always the same South America cards.

Then I've tested the client's modified version. I run a game with a bot and the GUI allows to make the reinforcements.

I'm attaching the patch as a unified patch file.^U
Ups, seems that GitHub doesn't allow me to attach files.
This is the patch content: http://pastebin.com/MGtQpuue
